promotion/drain.go

```diff
@@ -0,0 +1,92 @@
+package promotion
+
+import (
+	"context"
+
+	"github.com/brave-intl/bat-go/utils/clients/cbr"
+	"github.com/brave-intl/bat-go/wallet"
+	"github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+	"github.com/shopspring/decimal"
+)
+
+// Drain ad suggestions into verified wallet
+func (service *Service) Drain(ctx context.Context, credentials []CredentialBinding, walletID uuid.UUID) error {
+	wallet, err := service.datastore.GetWallet(walletID)
+	if err != nil || wallet == nil {
+		return errors.Wrap(err, "Error getting wallet")
+	}
+
+	// A verified wallet will have a payout address
+	if wallet.PayoutAddress == nil {
+		// Try to retrieve updated wallet from the ledger service
+		wallet, err = service.wallet.UpsertWallet(ctx, walletID)
+		if err != nil {
+			return errors.Wrap(err, "Error upserting wallet")
+		}
+
+		if wallet.PayoutAddress == nil {
+			return errors.New("Wallet is not verified")
+		}
+	}
+
+	// Iterate through each credential and assemble list of funding sources
+	_, _, fundingSources, err := service.GetCredentialRedemptions(ctx, credentials)
+	if err != nil {
+		return err
+	}
+
+	for _, v := range fundingSources {
+		if v.Type != "ads" {
+			return errors.New("Only ads suggestions can be drained")
```

Comment:
Since this is a loop, should invalid sources be skipped rather than failing everything?

Reply:
I think that is less safe then aborting. If in such a case we return success we run the risk of the client proceeding to delete tokens that the server was in fact unable to process ( unbeknownst to the client ). We could consider tracking and returning information about which funding sources were valid and able to be processed but then we're complicating the API and adding more surface area for issues.